src/infiniteLoops/detect.ts

```diff
@@ -0,0 +1,362 @@
+    circular = undefined
+  }
+  if (circular) {
+    let message
+    if (circular[0] === circular[1] && circular[0] === '') {
+      message = 'None of the variables are being updated.'
+    } else {
+      message = 'It has the infinite cycle: ' + circular.join(' -> ') + '.'
+    }
+    report(message, InfiniteLoopErrorType.Cycle)
+  } else {
+    const code = codeToDispatch(stackPositions, state)
+    const pass = runUntilValid(code)
+    if (pass) {
+      const message = 'In particular, ' + pass
+      report(message, InfiniteLoopErrorType.FromSmt)
+    }
+  }
+}
+
+/**
+ * If no if statement/conditional was encountered between iterations, there is no base case.
+ */
+function hasNoBaseCase(stackPositions: number[], state: st.State) {
+  const thePaths = state.mixedStack.slice(stackPositions[0], stackPositions[1]).map(x => x.paths)
+  return flatten(thePaths).length === 0
+}
+
+/**
+ * @returns if a cycle was detected, string array describing the cycle. Otherwise returns undefined.
+ */
+function checkForCycle(stackPositions: number[], state: st.State): string[] | undefined {
+  const maybeConc = stackPositions.map(i => state.getMaybeConc(i))
+  const concStr = []
+  for (const item of maybeConc) {
+    const innerStr = []
+    for (const [name, val] of item) {
+      if (typeof val === 'function') {
+        return
+      }
+      innerStr.push(`(${getOriginalName(name)}: ${stringifyCircular(val)})`)
+    }
+    concStr.push(innerStr.join(', '))
+  }
+  return getCycle(concStr)
+}
+
+function getCycle(temp: any[]) {
+  const last = temp[temp.length - 1]
+  const ix1 = temp.lastIndexOf(last, -2)
+  if (ix1 === -1) return undefined
+  const period = temp.length - ix1 - 1
+  const s1 = temp.slice(ix1 - period, ix1)
+  const s2 = temp.slice(ix1, -1)
+  if (s1.length != period) return undefined
+  for (let i = 0; i < period; i++) {
+    if (s1[i] != s2[i]) return undefined
+  }
+  return s1.concat(s1[0])
+}
+
+function stringifyCircular(x: any) {
+  // From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cyclic_object_value#examples
+  const getCircularReplacer = () => {
+    const seen = new WeakSet()
+    return (key: string, value: any) => {
+      if (typeof value === 'object' && value !== null) {
+        if (seen.has(value)) {
+          return '(CIRCULAR)'
+        }
+        seen.add(value)
+      }
+      return shallowConcretize(value)
+    }
+  }
+  return JSON.stringify(x, getCircularReplacer())
+}
+
+function runUntilValid(items: [string, () => string][]) {
+  for (const [code, message] of items) {
+    const out = runSMT(code)
+    if (out.includes('Valid')) return message()
+  }
+  return undefined
+}
+
+function runSMT(code: string): string {
+  try {
+    const input = JSON.stringify({ content: [code] })
+    const out = JSON.parse(runAltErgo(input, options))
+    return out.results[0]
+  } catch (e) {
+    return e.toString()
+  }
+}
+
+type Triple = [number[], number[], [string, any, number][]]
```

Comment:
Could you add a comment above this line describing the fields? Best case is to use an object here (but only if it doesn't further complicate the code).